doc/architecture.md

```diff
@@ -0,0 +1,80 @@
+
+The Rust compiler builds the Rust standard library as a static library and then links it into Rust code objects. The contents of this static library include the code which dynamically links against system interfaces. These system interfaces are what postgrestd intercepts by itself being a modification of the Rust standard library.
+
+The extension called "PL/Rust" which includes the language handler is responsible for covering the linking and loading steps. This extension may have privileges that user functions do not, using the Rust std of the same host target that PostgreSQL itself is compiled for, to interoperate in that privileged mode. This is as-usual for language handlers: they must typically be written in C.
+
+## Design Goals
+
+Design a custom rust compilation target for Postgres that provides nearly "safe" (as Rust defines it) and "trusted" (as Postgres defines a procedural language) plrust.
+
+The goals for the approach include
+
+* Architecture support for x86_64 and aarch64
+* Operating system support for Linux
+* Disallow File Handle operations
+* Disallow access to the internals of the database
+* Disallow access to the OS as the user executing the Postgres process 
+* Disallow access into active postmaster process, i.e. no ability to reach into Postgres memory space, despite executing inside it.
+* Gracefully handle rust panics and have them interoperate with Postgres' transaction system
+* Memory allocation within Postgres' palloc/pfree functions
+
+## Approach
+
+The plrust extension is compiled using the standard rust library.  The postgrestd library is used to compile the functions written using PLRust.  The postgrestd library is a libraries are compiled using The 
+
+Following an approach similar to the selection between libc and the musl libc standard library for compilation, a Postgres compilation target is defined that instructs the compiler to use the postgrestd library.  The postgrestd library provides the rust standard library interfaces except in the case where it is desirable to prevent access.  In those cases the code is [configured](https://doc.rust-lang.org/stable/rust-by-example/attribute/cfg.html) to be not present. The result is a small shim on top of the rust library limited access to the standard library.
+
+
+## Bird's Eye View
+
+![](assets/architecture_1.png)
+
+
+## Code Map
+
+This section talks briefly about various important directories and data structures.
```

Comment:
```suggestion
## Supporting Crates

Because PL/Rust implements a fairly complicated language and makes it sound to use as a trusted procedural language, there are multiple supporting crates necessary to make it work.

### pgx with `features = ["postgrestd", ..]`

The `plrust` language handler is itself implemented using pgx, but also all PL/Rust functions must understand the PostgreSQL data types in the same way that `pgx` enables. In addition, bindings via `pgx` are used to implement most of the other supporting crates.
```